### ISSUE TITLE
GPU training scripts in python send training infos on stderr instead of stdout

### ISSUE THREAD

**Issue by Potomac** [kaldi-asr/kaldi#1820]
Hello,

I use the very last version of kaldi,

in teldium/s5_r2 for the python version of train.sh ( train_dnn.py, train_rnn.py ) I notice that the output related to the progress of the GPU training is sent to stderr channel instead of stdout channel,

in my work we use a cluster,  log files ( stdout for normal output and stderr for error messages ) are created for each process, when I submit a kaldi job all is ok, but the output of scripts for the GPU training is redirected to stderr instead of stdout,

the previous version of kaldi scripts ( the bash version : training.sh ) didn't have this behaviour, stderr was used only for printing errors or warnings,

in the last version of tdnn/blstm/lstm scripts the stdout is done like this : 

"- informations about creation of run_ivectors_common"
 and it stops with after this steps : 
steps/nnet3/get_egs.sh: removing temporary alignments and transforms
steps/nnet3/get_egs.sh: Finished preparing training examples"

all the futur output is redirected to stderr : 
"- tree-info exp/mer3/tri4/tree 
- steps/nnet3/xconfig_to_configs.py --xconfig-file exp/nnet3_cleaned/tdnn1b_sp/configs/network.xconfig --config-dir exp/nnet3_cleaned/tdnn1b_sp/configs/
- 2017-08-13 12:46:11,346 [steps/libs/nnet3/train/frame_level_objf/common.py:191 - train_one_iteration - INFO ] Training neural net (pass 152)"

and the stdout resumes only at the end of the GPU training with a statistic report  : 

"exp/nnet3_cleaned/tdnn1b_sp: num-iters=156 nj=6..6 num-params=14.5M dim=40+100->9057 combine=-1.34->-1.33 loglike:train/valid[103,155,combined]=(-1.43,-1.35,-1.34/-1.84,-1.85,-1.84) accuracy:train/valid[103,155,combined]=(0.61,0.63,0.63/0.54,0.54,0.54)"

is this behaviour is normal, or is it a bug related to python ?

thanks

**Comment by danpovey** [comment]
It's not a bug.  You should merge the output and error streams for easier viewing.

**Comment by Potomac** [comment]
ok, thanks for the answer,
so I assume that a kind of "print >> sys.stderr" is currently used in these python scripts,

but it seems a weird way to manage the output,  by convention stderr channel  should be used only  for printing errors and warnings, 

it would be better to redirect the verbose information in a specific log, or give the possibility to disable the GPU training logging output with a switch --no-verbose-output

**Comment by danpovey** [comment]
Many programs print human-readable output to stderr which is not
necessarily error-related.  It's required any time a pipe may be involved,
which is the case for essentially all command-line Kaldi programs.

**Comment by vonmike** [comment]
If you are using default kaldi logging, it is possible to change the log behaviour by fixing the MessageLogger::LogMessage() function in kaldi-error.cc file:

```
if (envelope_.severity < LogMessageEnvelope::kWarning) {
    std::cerr << full_message.str();
} else {
    std::cout << full_message.str();
}
```

**Comment by danpovey** [comment]
No because Kaldi uses stdout for actual output as part of pipes.
